### PR TITLE
New EDS background 

### PR DESCRIPTION
**Pull request by ZanettaPM**
### Description of the change
This is an updated version of PR #1835 . There is not so many changes. It is now based on 1.6.4 

A new background component taking into account the detector efficiency, the composition of the sample and the edge jump is added to the components1D as an alternative to the Polynomial background.
This Background is based on Kramer emission law [Kramer 1923] (other emission models are also available) and convoluted with absorption model from Love and Scott [Sewell. et al, 1985] and the detector efficiency curve of the  SEM/TEM.

Key words: Kramer's Law, Mass absorption mixture function, quantification, better fit

### Progress of the PR
- [x] The component can be added to a model
- [x] The component is fitted with other components (gaussians)
- [x] The fit works with datasets with 1,2 or 3 dimensions
- [ ] At some point I would like to merge #2535 and use it to calculate the detector efficiency curve
- [ ] Need tests, but WIP
- [x] ready for new reviews.

### Example of how to fit the component

SEM example

```python

s=hs.datasets.example_signals.EDS_SEM_Spectrum()
s.add_lines()

m=s.create_model(auto_background=False)
m.add_physical_background(emission_model='Lifshin',correct_for_backscatterring=False)
m.components.Bremsstrahlung.initialize() 
m.fit_background(bounded=True)
m.fit()
m.plot(plot_components=True)

```

TEM example

```python
s = hs.datasets.example_signals.EDS_TEM_Spectrum()
s=s.isig[1.18:]
s.add_elements(['Ga','Cu','Si'])
s.add_lines()

kfactors = [1.77,1.450226,1,5.075602,1] #For Fe Ka and Pt La
bw = s.estimate_background_windows(line_width=[5.0, 2.0])
intensities = s.get_lines_intensity(background_windows=bw)
Weight_percent = s.quantification(intensities, method='CL', factors=kfactors,composition_units='weight' )

m=s.create_model(auto_background=False)
m.add_physical_background(quantification=Weight_percent) #here the detector is not the good one which explain the deviation at low energy
m.components.Bremsstrahlung.initialize()
m.fit_background(bounded=True) 
m.fit(bounded=True)
m.plot(plot_components=True)
```

### CODE REVIEW

File: hyperspy/misc/eds/detector_efficiency.py [hunk 1]
```diff
@@ -0,0 +1,1383 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with  HyperSpy.  If not, see <http://www.gnu.org/licenses/>.
+
+import numpy as np
+
+detector_efficiency ={}
+detector_efficiency['Polymer_C'] = ([[  0.10032033,   0.11797924],
```
Comment (ericpre) [inline hyperspy/misc/eds/detector_efficiency.py:22]: This need to be documented where does it come from.

File: doc/user_guide/eds.rst [hunk 2]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
 * :py:meth:`~.models.edsmodel.EDSModel.free_xray_lines_width`
 * :py:meth:`~.models.edsmodel.EDSModel.fix_xray_lines_width`
 
+.. _eds_background-Bremsstrahlung-label:
+
+Bremsstrahlung
```
Comment (ericpre) [inline doc/user_guide/eds.rst:656]: As @thomasaarholt already mentioned, there are numerous formatting issues, you can see the doc build of this PR at https://hyperspy--2836.org.readthedocs.build/en/2836/user_guide/eds.html#bremsstrahlung. The link is available at the bottom of this page in at the very bottom of the "checks" list - look for the "docs/readthedocs.org:hyperspy" build.

For more details on how to build the documentation locally, there are explanation in https://hyperspy.org/hyperspy-doc/current/dev_guide/writing_docs.html#user-guide-documentation

File: hyperspy/models/edsmodel.py [hunk 3]
```diff
@@ -301,12 +301,85 @@ def add_polynomial_background(self, order=6):
         self.append(background)
         self.background_components.append(background)
 
+    def add_physical_background(self, E0='from_metadata', detector='Polymer_C', quantification='Mean',emission_model='Kramer', absorption_model='quadrilateral', coating_thickness=0, TOA='from_metadata', Phase_map=None, correct_for_backscatterring=False, standard=False):
+        """
+        Add a physical model of the background taking into account the interraction e-/mater (see Zanetta et al. 2019)
```
Comment (ericpre) [inline hyperspy/models/edsmodel.py:306]: The reference itself needs to be added in docstring: https://numpydoc.readthedocs.io/en/latest/format.html#references

File: doc/user_guide/eds.rst (hunk 2 above)
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:656]: This is a big enough section that I think you can use the "subsection" header for this title - place "-----" under the title using the exact same number of symbols as the title has, e.g.:
```rst
Bremsstrahlung
--------------
```

File: doc/user_guide/eds.rst [hunk 4]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. versionadded:: 1.3
+
+a) emission
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:661]: Now that the above header uses subsection, use subsubsection `^^^^` here, remove the "a)" and capitalize?

File: doc/user_guide/eds.rst [hunk 5]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+Energetic electrons in matter create X-rays through one of two processes characteristic X-ray generation and bremsstrahlung. The x-ray bremsstrahlung is radiation generated as a result of the deceleration of beam electrons in the Coulombic fields of the target atoms. This radiation forms a slowly varying x-ray background ranging in energy from the incident electron energy down to zero. The intensity of this background is a function of the atomic number of the target Z, the energy of the bremsstrahlung x rays E, and the energy of the incident electrons E0. 
+In 1923 Kramers calculated a relationship for the bremsstrahlung emission based on electron scattering cross sections and the Thomas-Whiddington law for electron retardation in various targets. 
+
+N(E)=KZ ((E-E0)/E)
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:666]: You can use latex markup within rst, see [here](https://www.sphinx-doc.org/en/1.0/ext/math.html).
I suggest you do that, as it a) is much clearer what is multiplication of two symbols and what is a single symbol of two letters, and b) looks way nicer.

File: doc/user_guide/eds.rst [hunk 6]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+where N(E) is the number of bremsstrahlung photons of energy E (i.e., the intensity) produced by electrons of energy E0, K is Kramers’ constant, and Z is the atomic number of the ionized atom.
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:668]: You can try using the inline math markup to render the symbols to fit with the equation as well.

File: doc/user_guide/eds.rst [hunk 7]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+.. figure::  images/Bremsstrahlung_explanation.png
+   :align:   center
+   :width:   100
+
+    Absorption of the emitted Bremsstrahlung
+
+Other models of emission are available for this function (see Small et al. 1987, Castellano et l. 2004)
+
+b) Absorption in the sample
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:680]: subsubsection here as well

File: doc/user_guide/eds.rst [hunk 8]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+The absorption correction is based upon a simplified representation of the x-ray depth distribution known as Love & Scott model. A quadrilateral profile is assumed for the distribution of x-ray generation. This simplification has been detailed in (Sewell et al 1985). This model has shown a good robustness and seems to be adapted for the quantitative analysis of both heavy and light elements.
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:682]: 1. Capitalize X-ray
2. if possible include link to papers

File: doc/user_guide/eds.rst [hunk 9]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+Ac=((1-e^((-2µρx) ))/(2µρx))
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:684]: Use rst latex for this equation as well

File: doc/user_guide/eds.rst [hunk 10]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+The mass absorption coefficient is tabulated on an elemental basis, and mass absorption coefficients for compounds may be computed according to the following rule,
+
+µ=∑(Wi.µi)
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:688]: equation here as well

File: doc/user_guide/eds.rst [hunk 11]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+where Wi and µi are, respectively, the weight fraction and mass attenuation coefficient for element i in a compound with n elements.
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:690]: inline math symbols here :)

File: doc/user_guide/eds.rst [hunk 12]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+   The Mu coefficient calculated for all energy for a FeS sample
+
+.. figure::  images/Example_absorption.png
+   :align:   center
+   :width:   100
+
+   The corresponding absorption due to the sample
+
+c)	Detector efficiency correction
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:704]: subsubsection here

File: doc/user_guide/eds.rst [hunk 13]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+The second correction which was applied to the experimentally measured data involves an adjustment for the efficiency of the detector as a function of energy. 
+A fraction of the X-rays are absorbed by the thin window in front of the detector are before they reach the active measurement volume.
+
+Windows can be modeled using equation when the composition and thickness of the window are known. However, those data are not always available. The detector efficiency of different model of detectors have been added to hyperspy and can be selected in 'add_physical_background'
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:709]: Full stop at end

File: doc/user_guide/eds.rst [hunk 14]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+The second correction which was applied to the experimentally measured data involves an adjustment for the efficiency of the detector as a function of energy. 
+A fraction of the X-rays are absorbed by the thin window in front of the detector are before they reach the active measurement volume.
+
+Windows can be modeled using equation when the composition and thickness of the window are known. However, those data are not always available. The detector efficiency of different model of detectors have been added to hyperspy and can be selected in 'add_physical_background'  
+
+.. figure::  images/Example_polymer_window.png
+   :align:   center
+   :width:   100
+
+   example of the absorption effect though a polymer window
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:715]: Capitalize caption

File: doc/user_guide/eds.rst [hunk 15]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+d) Fitted expression
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:717]: subsubsection

File: doc/user_guide/eds.rst [hunk 16]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+Taking all those conditions in consideration only two free parameters where identified. The “KZ” parameter presented in the emission law, and the “ρx” parameter also known as mass depth. Resulting to a simple equation: 
+
+B(E)=a ((E-E0)/E)*((1-e^((-2µb) ))/(2µb))*Detector_eff
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:721]: Equation

File: doc/user_guide/eds.rst [hunk 17]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+.. figure::  images/Example_polymer_window.png
+   :align:   center
+   :width:   100
+
+   example of the absorption effect though a polymer window
+
+d) Fitted expression
+
+Taking all those conditions in consideration only two free parameters where identified. The “KZ” parameter presented in the emission law, and the “ρx” parameter also known as mass depth. Resulting to a simple equation:
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:719]: Inline math

File: doc/user_guide/eds.rst [hunk 18]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+   example of the absorption effect though a polymer window
+
+d) Fitted expression
+
+Taking all those conditions in consideration only two free parameters where identified. The “KZ” parameter presented in the emission law, and the “ρx” parameter also known as mass depth. Resulting to a simple equation: 
+
+B(E)=a ((E-E0)/E)*((1-e^((-2µb) ))/(2µb))*Detector_eff
+
+a and b are named coefficients and can be fixed or not in the model (notably with the function :py:meth:`~.fix_background`).
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:723]: Inline math

File: doc/user_guide/eds.rst [hunk 19]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+B(E)=a ((E-E0)/E)*((1-e^((-2µb) ))/(2µb))*Detector_eff
+
+a and b are named coefficients and can be fixed or not in the model (notably with the function :py:meth:`~.fix_background`). 
+
+.. figure::  images/Example_bremsstrahlung.png
+   :align:   center
+   :width:   100
+
+e) Example
```
Comment (thomasaarholt) [inline doc/user_guide/eds.rst:729]: Subsubection

**Review by thomasaarholt — COMMENTED** [review]
I added some grammar / rst suggestions. Also notice that github actions is mentioning some markup warnings - sometimes rst will silently hide sections completely if the markup has an error, so do compare the final view with the raw code afterwards.

Will be great to merge this!

File: hyperspy/models/edsmodel.py [hunk 20]
```diff
@@ -301,12 +301,85 @@ def add_polynomial_background(self, order=6):
...
+            and is only calculated one time for the entire phase
+        Correct_for_backscattering: True or False
+            apply a correction for the backscattered electron to the emitted intensities
+            
+             
+        Caution ! : The number of elements have to be equal to the number of Xray_lines. It's preferable to remove secondary lines and keep only higher energy lines
+        Example:
+            s = hs.datasets.example_signals.EDS_SEM_Spectrum()
+            s.add_lines()
+            m=s.create_model(auto_background=False)
```
Comment (thomasaarholt) [inline hyperspy/models/edsmodel.py:352]: add a space before and after the =

File: doc/user_guide/eds.rst [hunk 21]
```diff
@@ -651,6 +651,174 @@ functionalities of X-ray lines when fitting:
...
+
+Ac=((1-e^((-2µρx) ))/(2µρx))
+
+The mass absorption coefficient is tabulated on an elemental basis, and mass absorption coefficients for compounds may be computed according to the following rule,
+
+µ=∑(Wi.µi)
+
+where Wi and µi are, respectively, the weight fraction and mass attenuation coefficient for element i in a compound with n elements.
+
+.. figure::  images/Mucoefficient.png
```
Comment (ericpre) [inline doc/user_guide/eds.rst:692]: The images are too small. Where does the images come from? Can you add the code to generate the images?

**Review by ericpre — COMMENTED** [review]
Thanks for coming back on this @ZanettaPM! The git history wasn't in very good shape in #1835, and this is clearly much better in this PR.

The text in the user guide is good and the code seems to be in good shape. I will have a look at it in more details and add some more comments.
This is a nice addition and I can't agree more with @thomasaarholt that it would be great to have it merged! :)

It will need some tests of course!

**Comment by ericpre** [comment]
In #1835, there was some discussion on detector efficiency, which could be done in a separate PR - for example continuation of #2535.

**Comment by ZanettaPM** [comment]
@thomasaarholt, @ericpre, thanks a lot for your comments, after I updated the PR I indeed mentioned some rst typos, I will take care of it in the coming days. I will also add the references and clean a bit the documentation in general, some paragraph are some time too long. 
We can close #1835 if you want to decrease the list of open PRs.
Thanks again, more in the coming days ;)

**Comment by ZanettaPM** [comment]
> In #1835, there was some discussion on detector efficiency, which could be done in a separate PR - for example continuation of #2535.

Yes I updated the description of the PR. I can easily use the detector efficiency calculation #2535 in the background function. The question is, should I wait for #2535 and #2836 to be merged in the master code before adding this functionality?

In any case, numerous constructors do not always give the layer composition and thickness of the detectors, that is why I added a dictionnary containing some of the most common detector efficiency curves (I need to add the refs though ;) )

**Comment by ZanettaPM** [comment]
@CorentinLG also asked for the possibility to save the parameters of the background model. This can be tricky, there is many information by pixel that I do not save to not take to much memory, typically the Mu coefficient curve. Saving the final model is possible, some parameters too, but all of them might be useless and consume too much memory. (I just wanted to keep this note for later).

**Comment by ericpre** [comment]
Ping @ZanettaPM, so it doesn't get forgotten! :)

**Comment by ZanettaPM** [comment]
> Ping @ZanettaPM, so it doesn't get forgotten! :)

Haha, no it is still in my mind. I making small progress on it. 
Thanks !

**Comment by ericpre** [comment]
From @OliDG in #2989:

> Describe the bug
in Hyperspy/tree/bremsstrahlung_rebased/
Using "m_b.add_physical_background(detector='Super_X') " as in the docstring raises an error: KeyError: Super_X
>
> Solution:
in Hyperspy/tree/bremsstrahlung_rebased/hyperspy/misc/eds/detector_efficiency.py
>
> there is a typo l.457 :
> detector_efficiency[''SuperX']
> Should be: detector_efficiency['Super_X']
> to match the docstring of the instance and the naming style of the others.

**Comment by OliDG** [comment]
Is it a known issue? do you observe the same behavior? when using noisier data, the Bremsstrahlung background drop to zero?! 
Here is the same dataset with binning 3, binning 9 and the entire dataset.sum(); The background only seems to work for the high binning and sum() spectrum... I used multifit options for background and model: 

Physical background_Bremsstrahlung drop with noise

**Comment by ZanettaPM** [comment]
> Is it a known issue? do you observe the same behavior? when using noisier data, the Bremsstrahlung background drop to zero?! Here is the same dataset with binning 3, binning 9 and the entire dataset.sum(); The background only seems to work for the high binning and sum() spectrum... I used multifit options for background and model:
> 
> Physical background_Bremsstrahlung drop with noise

Yes, I am aware of the issue, you can try maybe to change the emission model, and see if you have better results. However, the number of counts in the bin 3 image alternate between 0 and 1 or 2, it is likely that the residuals are better if the background is set to 0 for this amount of counts. 

I can take a look into this but I am not sure if this is a coding problem or rather the model itself that might not be adapted for such noisy data. 

If you have a better model to put in equation, we could work on implementing it on your developer version.

**Comment by OliDG** [comment]
Thanks for the feedback. I see warnings about dividing by 0. I see 2 ways out, either there is a way to ignore the 0 values to fit the model (and check the residual), or one can add +1 to each pixel to avoid 'Nan' values, I assume this addition will be compensated by a slightly increased background level (by 1 unit) and would allow avoiding all the Nan errors. I will try to look for other models too when I find the time :D

Edit: the +1 to the dataset allows the background to be fitted at most pixels but seems to alter the fitting quality, more specifically, the background drops to 0 for higher kV and will not compensate for the +1 as I expected.

**Comment by thomasaarholt** [comment]
It won’t help with your edit (I think) but could you try adding a much smaller value, like 0.0001?

**Comment by OliDG** [comment]
> 

it is a better idea thanks, at least the data is barely altered. It seems to work as well, unfortunately, as for the +1, there are still 20% of the pixels with a background at 0.

**Comment by thomasaarholt** [comment]
I would run PCA on the whole dataset, keeping a high number of components (at least 30, maybe a hundred). Then fit again.

**Comment by adriente** [comment]
I had a discussion with Corentin Leguillou about the background issue. 

I think it is a fitting method issue. The usual method is based on the least-square loss function which does not work very well for data that follow a Poisson statistical law and especially in the case of low number of counts per channel. 

You can enable the loss-function keyword argument of the fit function and change it to "ML-Poisson". There is a little bit more of information in the documentation of hyperspy : http://hyperspy.org/hyperspy-doc/current/user_guide/model.html
I am not sure but I expect the fit function to work better with a better loss function.

PCA is probably not going to help as it shares the same kind of issues with very noisy data for the somewhat the same reasons.
I am happy to discuss it further but probably not here to avoid spamming.

**Comment by jlaehne** [comment]
Replaced by #3044

**Pull request closed without merging**